### PR TITLE
Add a request pool

### LINKED ISSUE #1100: Provide option to limit requests per second
## Description

Currently there can be > 1000 API requests per second depending on how the accounts look and what options are used, this can lead to some errors because it exceeds rate limits from the server side.
Maybe we can add a global rate limit using a crate like https://crates.io/crates/governor? Or are there other options?
We also limit our parallel requests in some places with https://github.com/iotaledger/iota-sdk/blob/4bf72ecb13c3540895d03cc62eb527241e41675a/sdk/src/client/constants.rs#L24, could this then be removed, or should we also make it modifiable by the user?

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

This PR adds a request pool to the client, which simply waits for a limited resource to arrive on a channel before calling any client request. This is implemented using an extension trait for `Future` which first waits for permission from the pool to make the request, then calls the original future, then returns the permission to the pool.

## Links to any relevant issues

Closes #1100 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Bug fix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Changelog worthy

File: sdk/src/client/node_manager/mod.rs [hunk 1]
```diff
@@ -58,6 +63,104 @@ impl std::fmt::Debug for NodeManager {
     }
 }
 
+impl ClientInner {
+    pub(crate) async fn get_request<T: DeserializeOwned + Debug + Serialize>(
+        &self,
+        path: &str,
+        query: Option<&str>,
+        timeout: Duration,
```
Comment (thibault-martinez) [inline sdk/src/client/node_manager/mod.rs]: I wonder if these new methods need a timeout since we have self.timeout

Comment (thibault-martinez) [reply sdk/src/client/node_manager/mod.rs]: Maybe at least optional if not removed?

File: sdk/src/client/node_api/core/mod.rs [hunk 2]
```diff
@@ -15,87 +13,29 @@ use crate::{
...
-                    .await?
-                }
-            }))
-            .await?
-            .into_iter()
-            .flatten()
-            .collect();
-
-        Ok(outputs)
+        futures::future::try_join_all(output_ids.iter().map(|id| self.get_output(id))).await
```
Comment (thibault-martinez) [inline sdk/src/client/node_api/core/mod.rs:16]: Can you explain this change please?

Comment (DaughterOfMars) [reply sdk/src/client/node_api/core/mod.rs:16]: We just don't need to do any special handling here anymore

File: sdk/src/client/node_manager/mod.rs (hunk 1 above)
Comment (DaughterOfMars) [reply sdk/src/client/node_manager/mod.rs]: made it better

File: sdk/src/client/node_manager/mod.rs [hunk 3]
```diff
@@ -58,6 +63,73 @@ impl std::fmt::Debug for NodeManager {
...
 
+impl ClientInner {
+    pub(crate) async fn get_request<T: DeserializeOwned + Debug + Serialize>(
+        &self,
+        path: &str,
+        query: Option<&str>,
+        need_quorum: bool,
+        prefer_permanode: bool,
+    ) -> Result<T> {
+        #[cfg(not(target_family = "wasm"))]
```
Comment (kwek20) [inline sdk/src/client/node_manager/mod.rs:76]: for all of these, we could make it codewise non-duplicative spacing the request and rate limit behind the #not(wasm).
Better for maintenance id say

**Review by Thoralf-M — CHANGES_REQUESTED** [review]
Just blocks forever when amount of parallel requests is > max_parallel_api_requests

Build with `.with_max_parallel_api_requests(2)` and then sync an account

```
2023-09-06 09:12:34 (UTC) iota_sdk::wallet::account::operations::syncing::addresses::output_ids::alias_foundry DEBUG [SYNC] get_alias_and_foundry_output_ids
2023-09-06 09:12:34 (UTC) reqwest::connect                           DEBUG starting new connection: https://api.testnet.shimmer.network/
2023-09-06 09:12:34 (UTC) reqwest::connect                           DEBUG starting new connection: https://api.testnet.shimmer.network/
2023-09-06 09:12:34 (UTC) iota_sdk::client::node_manager::http_client DEBUG GET: 186 ms for 200 OK https://api.testnet.shimmer.network/api/indexer/v1/outputs/basic?expirationReturnAddress=rms1qqhfjcm5gg77l9m6lmqsgr42c5drwtt9dexhpuky0kd8fhccvg8uj9k9l7h
2023-09-06 09:12:34 (UTC) iota_sdk::client::node_manager::http_client DEBUG GET: 188 ms for 200 OK https://api.testnet.shimmer.network/api/indexer/v1/outputs/basic?address=rms1qqhfjcm5gg77l9m6lmqsgr42c5drwtt9dexhpuky0kd8fhccvg8uj9k9l7h
```
and then nothing happens anymore (after some time the node syncing request, but syncing never continues)


Full example to reproduce
```Rust
use iota_sdk::{
    client::{
        constants::SHIMMER_COIN_TYPE,
        secret::{mnemonic::MnemonicSecretManager, SecretManager},
    },
    wallet::{ClientOptions, Result, Wallet},
};

#[tokio::main]
async fn main() -> Result<()> {
    dotenvy::dotenv().ok();

    let logger_output_config = fern_logger::LoggerOutputConfigBuilder::new()
        .name("example.log")
        .target_exclusions(&["h2", "hyper", "rustls"])
        .level_filter(log::LevelFilter::Debug);
    let config = fern_logger::LoggerConfig::build()
        .with_output(logger_output_config)
        .finish();
    fern_logger::logger_init(config).unwrap();

    let client_options = ClientOptions::new()
        .with_node(&std::env::var("NODE_URL").unwrap())?
        .with_max_parallel_api_requests(2);
    let secret_manager = MnemonicSecretManager::try_from_mnemonic(iota_sdk::client::generate_mnemonic()?)?;
    let wallet = Wallet::builder()
        .with_secret_manager(SecretManager::Mnemonic(secret_manager))
        .with_client_options(client_options)
        .with_coin_type(SHIMMER_COIN_TYPE)
        .finish()
        .await?;

    let account = wallet.create_account().with_alias("Alice").finish().await?;

    println!("Syncing account");
    account.sync(None).await?;
    Ok(())
}
```

**Review by Thoralf-M — CHANGES_REQUESTED** [review]
Works now, but please also add maxParallelApiRequests to the bindings ClientOptions

File: sdk/src/client/node_manager/mod.rs (hunk 3 above)
Comment (DaughterOfMars) [reply sdk/src/client/node_manager/mod.rs:76]: done

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/client/client-options.ts [hunk 4]
```diff
@@ -36,6 +36,8 @@ export interface IClientOptions {
     powWorkerCount?: number;
     /** Whether the PoW should be done locally or remotely. */
     localPow?: boolean;
+    /** The maximum parallel API requests. */
+    maxParallelApiRequests?: number;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/client/client-options.ts:40]: Changelog

File: bindings/python/iota_sdk/types/client_options.py [hunk 5]
```diff
@@ -103,6 +105,7 @@ class ClientOptions:
     apiTimeout: Optional[Duration] = None
     remotePowTimeout: Optional[Duration] = None
     powWorkerCount: Optional[int] = None
+    maxParallelApiRequests: Optional[int] = None
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/client_options.py:108]: Changelog

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/CHANGELOG.md [hunk 6]
```diff
@@ -27,6 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `Wallet::get_chrysalis_data()` method;
 - `PrivateKeySecretManager` and `SecretManager::PrivateKey`;
 - `SecretManager::from` impl for variants;
+- `Client` requests now obey a maximum concurrency using a request pool (`set via ClientBuilder::with_max_parallel_api_requests`);
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md]:
```suggestion
- `Client` requests now obey a maximum concurrency using a request pool (set via `ClientBuilder::with_max_parallel_api_requests`);
```

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/client/request_pool.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    pub(crate) fn new(size: usize) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(RequestPoolInner::new(size))),
+        }
+    }
+
+    pub(crate) async fn borrow(&self) -> Requester {
+        // Get permission to request
+        let mut lock = self.write().await;
+        lock.recv.recv().await;
```
Comment (thibault-martinez) [inline sdk/src/client/request_pool.rs:40]: What's this for exactly?

Comment (thibault-martinez) [reply sdk/src/client/request_pool.rs:40]: Ah it makes sense now that I read the Drop impl

**Review by thibault-martinez — DISMISSED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**